### PR TITLE
Remove CustomLambdaOutgoing

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR streamlines the code generator for proxies and removes CustomLambdaOutgoing.

### CODE REVIEW

File: cpp/include/Ice/OutgoingAsync.h [hunk 1]
```diff
@@ -407,9 +407,13 @@ class LambdaOutgoing : public OutgoingAsyncT<R>, public LambdaInvoke
                 {
                     response(std::move(v));
                 }
-                catch(...)
+                catch (const std::exception& ex)
```
Comment (bernardnormier) [inline cpp/include/Ice/OutgoingAsync.h:410]: We no longer throw/catch exception_ptr anywhere.

File: cpp/src/slice2cpp/Gen.cpp [hunk 2]
```diff
@@ -543,45 +543,70 @@ writeOpDocSummary(Output& out, const OperationPtr& op, const CommentPtr& doc, Op
...
+        {
+            out << element;
+        }
+        out << eabrk;
+        return os.str();
+    }
+}
 
+vector<string>
+createOutgoingAsyncParams(const OperationPtr& p, const string& scope, int typeContext)
```
Comment (externl) [inline cpp/src/slice2cpp/Gen.cpp:571]: Would be nice to have a comment here. It sounds similar to the one above it, but takes a very different parameters.

File: cpp/src/slice2cpp/Gen.cpp [hunk 3]
```diff
@@ -1826,167 +1822,120 @@ Slice::Gen::ProxyVisitor::visitOperation(const OperationPtr& p)
...
     }
 
-    C << "::std::function<void " << spar << lambdaOutParams << epar << "> response,";
-    C << nl << "::std::function<void(::std::exception_ptr)> ex,";
-    C << nl << "::std::function<void(bool)> sent,";
-    C << nl << "const ::Ice::Context& context) const";
-    C.restoreIndent();
+    C << nl << "return ::IceInternal::makeLambdaOutgoing<" << lambdaT << ">" << spar;
+
+    C << "std::move(" + (lambdaOutParams.size() > 1 ? string("responseCb") : "response") + ")"
```
Comment (externl) [inline cpp/src/slice2cpp/Gen.cpp:1860]: We should use the stream operation here.

**Review by externl — APPROVED** [review]
Looks good.

File: cpp/src/slice2cpp/Gen.cpp (hunk 2 above)
Comment (bernardnormier) [reply cpp/src/slice2cpp/Gen.cpp:571]: added comment.

File: cpp/src/slice2cpp/Gen.cpp (hunk 3 above)
Comment (bernardnormier) [reply cpp/src/slice2cpp/Gen.cpp:1860]: The rule is simple: within a spar/epar (and similar), using the stream operator adds a separator (comma with spar/epar). So when we don't want a comma, we must use "+".

Comment (externl) [inline cpp/src/slice2cpp/Gen.cpp:1860]:
```suggestion
    C << "std::move(" + (lambdaOutParams.size() > 1 ? "responseCb" : "response") + ")"
```

Comment (bernardnormier) [reply cpp/src/slice2cpp/Gen.cpp:1860]: It's required - doesn't compile without.

File: cpp/include/Ice/OutgoingAsync.h [hunk 4]
```diff
@@ -407,9 +407,13 @@ class LambdaOutgoing : public OutgoingAsyncT<R>, public LambdaInvoke
                 {
                     response(std::move(v));
                 }
-                catch(...)
+                catch (const std::exception& exception)
                 {
-                    throw std::current_exception();
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/include/Ice/OutgoingAsync.h:412]: Fixed as suggested.

File: cpp/src/slice2cpp/Gen.cpp [hunk 5]
```diff
@@ -543,45 +543,72 @@ writeOpDocSummary(Output& out, const OperationPtr& op, const CommentPtr& doc, Op
 
 // Returns the client-side result type for an operation - can be void, a single type, or a tuple.
 string
-createResultType(const OperationPtr& p, const string& scope, int useWstring)
+createOutgoingAsyncTypeParam(const vector<string>& elements)
 {
-    assert(useWstring == 0 || useWstring == TypeContextUseWstring);
+    if (elements.size() == 0)
```
Comment (pepone) [inline cpp/src/slice2cpp/Gen.cpp:548]: or `switch(elements.size())`

File: cpp/src/slice2cpp/Gen.cpp [hunk 6]
```diff
@@ -543,45 +543,72 @@ writeOpDocSummary(Output& out, const OperationPtr& op, const CommentPtr& doc, Op
...
+        }
+        out << eabrk;
+        return os.str();
+    }
+}
 
+// Returns the C++ types that make up the client-side result type of an operation (first return type then out parameter
+// types, as per the future API).
+vector<string>
+createOutgoingAsyncParams(const OperationPtr& p, const string& scope, int typeContext)
```
Comment (pepone) [inline cpp/src/slice2cpp/Gen.cpp:573]: Can we simplify this method like:

```
vector<string>
createOutgoingAsyncParams(const OperationPtr& p, const string& scope, int typeContext)
{
    vector<string> elements;
    TypePtr ret = p->returnType();
    if (ret)
    {
        elements.push_back(returnTypeToString(ret, p->returnIsOptional(), scope, p->getMetaData(), typeContext));
    }
    for (const auto& param : p->outParameters())
    {
        elements.push_back(typeToString(param->type(), param->optional(), scope, param->getMetaData(), typeContext));
    }
    return elements;
}
```

File: cpp/src/slice2cpp/Gen.cpp [hunk 7]
```diff
@@ -1627,77 +1654,52 @@ Slice::Gen::ProxyVisitor::visitOperation(const OperationPtr& p)
...
+    vector<string> lambdaOutParams = createOutgoingAsyncParams(p, interfaceScope, _useWstring | TypeContextInParam);
 
-    string returnValueS = "returnValue";
-    bool outParamsHasOpt = false;
+    string futureImplPrefix = "_iceI_";
+    string lambdaImplPrefix = "_iceI_";
 
-    if(ret)
+    if (futureOutParams != lambdaOutParams)
     {
```
Comment (pepone) [inline cpp/src/slice2cpp/Gen.cpp:1673]: What about

```
@@ -1666,13 +1649,8 @@ Slice::Gen::ProxyVisitor::visitOperation(const OperationPtr& p)
     vector<string> futureOutParams = createOutgoingAsyncParams(p, interfaceScope, _useWstring);
     vector<string> lambdaOutParams = createOutgoingAsyncParams(p, interfaceScope, _useWstring | TypeContextInParam);
 
-    string futureImplPrefix = "_iceI_";
-    string lambdaImplPrefix = "_iceI_";
-
-    if (futureOutParams != lambdaOutParams)
-    {
-        lambdaImplPrefix = "_iceIL_";
-    }
+    const string futureImplPrefix = "_iceI_";
+    const string lambdaImplPrefix = futureOutParams == lambdaOutParams ? "_iceI_" : "_iceIL_";
```

**Review by pepone — APPROVED** [review]

File: cpp/src/slice2cpp/Gen.cpp (hunk 6 above)
Comment (bernardnormier) [reply cpp/src/slice2cpp/Gen.cpp:573]: Fixed as suggested.

File: cpp/src/slice2cpp/Gen.cpp (hunk 7 above)
Comment (bernardnormier) [reply cpp/src/slice2cpp/Gen.cpp:1673]: Simplified as suggested.

**Pull request merged**